sorting: its not working lol

ashlsun opened the pull request:
the placement of the sorting select input is just temporary - could imagine this just being moved into a settings box or be something like cickable table headings on a grid layout, like spotify

anyway i was just trying to see if i could get it to work but . i don't think i fully understand svelte reactivity cause i initially just tried to mutate `list` and changed it to this when it wasn't visibly updating anything. it still doesn't visibly update 😞 
this is the error i got
after switching to "a to z" <img width="1404" alt="Screenshot 2024-07-10 at 11 45 51 PM" src="https://github.com/ashlsun/untitled-inventory-app/assets/37150737/2f364b9a-5eab-4998-a933-f3de70b14521">

i wonder what a proxy array / object is & why its not just a regular list?
im also finding it interesting that this null error comes after sorting and assigning. is it happening somewhere in the components?

ashlsun committed:
its not working lol

Rettend committed:
refactor: sort type string to string literal

Rettend committed:
refactor!: global item store

Rettend committed:
refactor: group storage related ops

Rettend committed:
chore: cleanup

Rettend commented:
i did a major refactor and changed some behaviors a little:
- [x] adding an already existing item now increments the quantity
- [x] this also fixes adding random items when they are full
- [x] when selecting items with the up/down arrows, it now jumps between storages upon reaching the edge

ideas for further improvs:
- [ ] add a small delay when pressing down the up/down arrow keys, it currently navigates so fast you can't see it
- [x] use numbers and maybe ctrl to quickly navigate between storages. like ctrl + 1 to select the first item of the first storage, ctrl + 2 to select the second storage's first item (oh ctrl + a number already jumps between tabs in the browser so we would need different shortcuts or maybe override the broswer's?)
- [x] fix sorting aaaaaa

ashlsun commented:
noticing:
- [ ] if a user is visiting for the first time (if nothing is in the db) there are no storage places on the screen. i think it would be nice to display an empty fridge and empty freezer
- [x] item count doesn't update
- [x] sorting is sorta working?? just sorts according to the previously chosen sortBy
- [ ] previously, deleting the last item in a storage place set the new last item as the selected, which i think would be nice to keep!
- [x] previously, entering a new item will unselect any selected item in the storage place, now it stays selected

hmm and i am having trouble reproducing this, but there was a moment upon first loading when pressing enter would not add an item (the input text would disappear but the item would not be created)! i'll keep my eyes peeled if it happens again

ashlsun commented:
i love the wrapping to the next storage when reaching the edge tho 🤩

ashlsun committed:
move parsing logic to handler in component, remove importItem

ashlsun commented:
i thought it might make more sense to just have the parsing happen in StoragePlace instead of the store, which avoids the add/import naming confusion

Rettend committed:
deps: bump svelte and fix sorting bruhhhh

Rettend committed:
chore: resolve conflicts

Rettend committed:
fix: messed up format in tips

Rettend committed:
revert: key with id

ashlsun commented:
wondering why sorting is non responsive and then works after a little bit for me
https://github.com/user-attachments/assets/b23cc3da-2189-455d-96a3-0d60623b4c73

nvm!! might have been the item.name as id

ashlsun commented:
ok so i added an apple just now, and then i updated its dateAdded to be 14d ago. when sorting by newest it should be the middle item, but it seems like it still treats it as the newest
https://github.com/user-attachments/assets/d8159a9c-933d-42fd-91e5-f0c01a235295

might also make sense to start with `sort by: none` at first, and then if a sort by option is selected, any changes to any of the items in the relevant attributes might cause them to be resorted

Rettend committed:
feat: add none sort, remove hardcodes

Rettend commented:
> ok so i added an apple just now, and then i updated its dateAdded to be 14d ago. when sorting by newest it should be the middle item, but it seems like it still treats it as the newest https://github.com/user-attachments/assets/d8159a9c-933d-42fd-91e5-f0c01a235295
> 
> might also make sense to start with `sort by: none` at first, and then if a sort by option is selected, any changes to any of the items in the relevant attributes might cause them to be resorted

i could not reproduce that bug you had, but added the `'none'` sort option. when adding a new item it switches to none, but this does not work when adding random items (debug feature anyway)

Rettend committed:
feat: store sorts in localstorage

Rettend committed:
style: change delete text to x icon

Rettend commented:
how about this? 
![image](https://github.com/user-attachments/assets/47de47f0-67ff-4978-90a9-2b6aeefc5c63)

a text `x` wouldn't be bad either but cant really increase the font size, or the gap between them grows too:
![image](https://github.com/user-attachments/assets/d944f401-df0b-4051-bb8e-930b46241364)

Rettend committed:
fix: item count, tests broken by delete icon

* item count was not reactive, can't make $derived value with parameter so a record of itemCounts was created
* fixed: tests that used text="delete" broke after swapping it to an icon

Rettend committed:
fix: preserve selected item when sorting

Rettend committed:
feat: add and rename storages

Rettend commented:
- [ ] use flexbox on storages, row wrap for desktop. col for mobile

ashlsun committed:
make x icon a thinner stroke

Rettend commented on tailwind.config.js:
```diff
@@ -23,7 +23,10 @@ export default {
         switch (prefix) {
           case 'tabler':
-            return content.replaceAll('stroke-width="2"', 'stroke-width="3"')
+            if (name === 'x')
```
wow nice

ashlsun committed:
use option/alt + number to select the first item of corresponding storage

ashlsun commented on src/lib/stores/item.svelte.ts:
```diff
@@ -1,86 +1,198 @@
+      if (!storages.includes(storage)) {
+        storages.push(storage)
+        items[storage] = []
+        await db.foodItems.where('storage').equals(storage).toArray()
```
is the result of this supposed to be stored anywhere?

Rettend replied:
hm no that can be removed it seems, no need to query there. i'm still not 100% sure how dexie works

ashlsun committed:
don't select when storage is an empty list

Rettend committed:
fix: remove random idb query

Rettend committed:
feat: add flex wrapping to storages

* should work on mobile too

Rettend committed:
fix: prevent scrolling with arrows

* should not scroll when an item is selected

Rettend committed:
refactor: change sort dropdown handling for tests

Rettend committed:
test: add basic tests for storageplace

ashlsun commented:
huh! pulled the latest updates and encountered this bug where clicking on an item in the 1st storage place ended up selecting an item in 2nd one. after that error, clicking anything else stops working 😮 

https://github.com/user-attachments/assets/429ac201-dd4a-4a5c-aa67-6b89e863513d

i tried going back one commit at a time, and it seems like the bug doesnt happen until after c02e645, refactor: change sort dropdown handling for tests? checking out 8e6ac1b and running the app, clicking on items seems to work as expected

its the removal of untrack, isnt it?

Rettend committed:
fix: ♾️ recursion

Rettend commented:
> huh! pulled the latest updates and encountered this bug where clicking on an item in the 1st storage place ended up selecting an item in 2nd one. after that error, clicking anything else stops working 😮
> 
>  Screen.Recording.2024-07-17.at.10.30.20.PM.mov 
> i tried going back one commit at a time, and it seems like the bug doesnt happen until after [c02e645](https://github.com/ashlsun/untitled-inventory-app/commit/c02e645bddef1d27512820347c2e3c08249e400d), refactor: change sort dropdown handling for tests? checking out [8e6ac1b](https://github.com/ashlsun/untitled-inventory-app/commit/8e6ac1bf2cb3ef034e08c93254f49a10715b0363) and running the app, clicking on items seems to work as expected
> 
> its the removal of untrack, isnt it?

ah sryyyy, it was untrack.

ashlsun committed:
fix ai workflow

ashlsun committed:
clear selection after input

ashlsun committed:
fix: select previous item if deleting last item

ashlsun committed:
handle keyboard navigation at page level instead of item level

ashlsun committed:
fix tests & remove obsolete ones
